lock-api/src/main/java/com/palantir/lock/watch/ConjureLockRequestMetadataUtils.java

```diff
@@ -0,0 +1,132 @@
+import com.palantir.util.IndexEncodingUtils.KeyListChecksum;
+import com.palantir.util.Pair;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+public final class ConjureLockRequestMetadataUtils {
+    private ConjureLockRequestMetadataUtils() {}
+
+    @VisibleForTesting
+    static final ChecksumType DEFAULT_CHECKSUM_TYPE = ChecksumType.CRC32_OF_DETERMINISTIC_HASHCODE;
+
+    public static Pair<List<LockDescriptor>, ConjureLockRequestMetadata> toConjureIndexEncoded(
+            Set<LockDescriptor> lockDescriptors, LockRequestMetadata metadata) {
+        ChangeMetadataToConjureVisitor toConjureVisitor = new ChangeMetadataToConjureVisitor();
+        IndexEncodingResult<LockDescriptor, ConjureChangeMetadata> encoded = IndexEncodingUtils.encode(
+                lockDescriptors,
+                metadata.lockDescriptorToChangeMetadata(),
+                changeMetadata -> changeMetadata.accept(toConjureVisitor),
+                DEFAULT_CHECKSUM_TYPE);
+        KeyListChecksum checksum = encoded.keyListChecksum();
+        ConjureLockRequestMetadata conjureLockRequestMetadata = ConjureLockRequestMetadata.of(
+                encoded.indexToValue(), checksum.type().getId(), Bytes.from(checksum.value()));
+        return Pair.create(encoded.keyList(), conjureLockRequestMetadata);
+    }
+
+    public static LockRequestMetadata fromConjureIndexEncoded(
+            List<LockDescriptor> keyList, ConjureLockRequestMetadata conjureMetadata) {
+        ChangeMetadataFromConjureVisitor fromConjureVisitor = new ChangeMetadataFromConjureVisitor();
+        ChecksumType checksumType = ChecksumType.valueOf(conjureMetadata.getChecksumTypeId());
+        KeyListChecksum checksum = KeyListChecksum.of(
+                checksumType, conjureMetadata.getChecksumValue().asNewByteArray());
+        IndexEncodingResult<LockDescriptor, ConjureChangeMetadata> encoded =
+                IndexEncodingResult.of(keyList, conjureMetadata.getIndexToChangeMetadata(), checksum);
+        Map<LockDescriptor, ChangeMetadata> changeMetadata = IndexEncodingUtils.decode(
+                encoded, conjureChangeMetadata -> conjureChangeMetadata.accept(fromConjureVisitor));
+        // visitUnknown() will return null
+        changeMetadata.values().removeIf(Objects::isNull);
+        return LockRequestMetadata.of(changeMetadata);
+    }
+
+    private static final class ChangeMetadataToConjureVisitor implements ChangeMetadata.Visitor<ConjureChangeMetadata> {
+
+        @Override
+        public ConjureChangeMetadata visit(Unchanged unchanged) {
+            return ConjureChangeMetadata.unchanged(ConjureUnchangedChangeMetadata.of());
+        }
+
+        @Override
+        public ConjureChangeMetadata visit(Updated updated) {
+            return ConjureChangeMetadata.updated(
+                    ConjureUpdatedChangeMetadata.of(Bytes.from(updated.oldValue()), Bytes.from(updated.newValue())));
```

Comment:
there are two calls to copy the array here which annoys me a bit tbh. I do not see another way around this though

Reply:
Yeah `Bytes` doesn't give us an easy way around this, i this this is fine.